Pool model

bquorning opened the pull request:
TODO: Split into several pull requests.

bquorning committed:
Add Pool model

Move pool logic incl. building/caching out of Event and into a separate
file.

bquorning committed:
Use Event#sort_column for finding previous events

Now the previous events are always returned in predictable order, and it
is also possible to find the single previous event without instantiating
all previous events.

By the way, the old `this_and_previous_events[-2]` would not always
return the correct event. If multiple events are executed on the same
date, all those events would be part of `#this_and_previous_events`. If
e.g. 3 events were executed on that date, they would be on indexes -3,
-2 and -1. So it's easy to imagine `this_and_previous_events[-2]`
pointing not to the previous event, but in fact to "this" event. It
feels nice to finally have this bug fixed.

bquorning committed:
Build pool smarter

Build an event's pool based on just the previous event's pool, not *all*
previous events. Of course, if the pool are not in the cache, this code
will be called recursively through all previous events. But if it *is*
cached, it should be *much* faster than the old code.

bquorning committed:
Persist Pool data in database instead of memory cache

bquorning commented:
FYI @libo I’ve been tinkering with making a proper Pool model instead of using the cache. This makes it feasible to calculate the avg. carrying by just using numbers from the previous event’s pool instead of going through _all_ previous events each time we display an event whole pool isn’t in the cache.

The ordering of events needs to be consistent, and correct. I’ve added an `events.sort_column` column. It will be automatically created on all new (and the rare updated) events, and I added a Rake task for backfilling those values.

When adding/updating/deleting an event, that event’s pool and all pools after it get deleted and re-calculated. Because the pool calculation is recursive, I just ask for the pool for the most recent event to be created.

I have not yet looked at adding specs, or even fixing/removing the existing specs. I wanted to hear your input on this approach before proceeding any further.

bquorning committed:
Add a sort_column to Event

Now we have a way to consistently find the previous event.

The Rake task `tesoro:backfill_event_sort_column` needs to be run on all
existing databases.

bquorning committed:
Ugly hack to make specs pass 😅

libo commented on app/controllers/events_controller.rb:
```diff
@@ -49,13 +51,17 @@ def create
 
   def destroy
     @event.destroy
+    Pool.recalculate(@event)
```
We need also to handle the scenario where we update the conversion rates for USD.
In such case we need to wipe the pool too.

bquorning replied:
That is true. The whole cache invalidation thing is hard 😬

libo commented on db/migrate/20211231004018_add_sort_column_to_events.rb:
```diff
@@ -0,0 +1,7 @@
+class AddSortColumnToEvents < ActiveRecord::Migration[6.1]
```
Can we merge into one migration that adds the table with `sort_column` ?

libo replied:
Forget what I said

libo commented:
Wow the difference in term of perf is huge 10x

Before with a test of 290 events

<img width="1423" alt="Screenshot 2022-01-03 at 09 58 48" src="https://user-images.githubusercontent.com/37512/147913884-243186c2-952d-472d-a015-a16a4ef1c627.png">

After same test
<img width="1424" alt="Screenshot 2022-01-03 at 10 02 46" src="https://user-images.githubusercontent.com/37512/147913992-55653fef-13a3-46fe-831a-2c06598af4be.png">

libo commented on app/models/event.rb:
```diff
@@ -75,12 +61,23 @@ def conversion_rate
+
+  def set_sort_column
+    # Ordered so that buys are before sales.
+    self.sort_column = "#{executed_on.iso8601}/#{self.class.actions[action]}/#{Time.now.to_f}"
```
If we really want to sort buy events before the sell does the `Time.now.to_f` allows that?

libo replied:
Yes, sorting of strings is from left to right

bquorning replied:
The `Time.now.to_f` is just to have _something_ to separate the different events happening on the same date. I initially wanted to use just `#{id}`, but that of course fails on new records, because we don’t have the ID before the row is persisted.

libo commented on app/models/event.rb:
```diff
@@ -3,6 +3,8 @@ class Event < ApplicationRecord
   belongs_to :user
   belongs_to :currency
 
+  has_one :pool
```
`dependent: :destroy`?

bquorning replied:
Hmm, I’m not sure. When an event is destroyed we want more things to happen than just destroying the associated pool.

libo commented on app/models/pool.rb:
```diff
@@ -0,0 +1,53 @@
+class Pool < ApplicationRecord
+  belongs_to :event
```
I don't remember how but I managed by testing the app on this branch to create a pool with size 0.

We should add some validation to avoid this scenario.

```ruby
>> self.sell.events_for_year(2022).second.capital_gain
=> NaN
>> self.sell.events_for_year(2022).second
=> #<Event id: 2, action: "sell", user_id: 1, currency_id: 1, stock_id: 1, quantity: 1, price: 0.1e1, commission: 0.0, executed_on: "2022-01-15", created_at: "2022-01-03 09:22:02.558546000 +0000", updated_at: "2022-01-03 09:22:02.558546000 +0000", sort_column: "2022-01-15/1/1641201722.554402">
>> self.sell.events_for_year(2022).second.pool
=> #<Pool id: 297, event_id: 2, value: 0.0, size: 0>
```

bquorning replied:
This would happen if `previous_event.pool.size` is zero. But how did you end up with an empty pool for two events in a row? The last event sells 1 stock, so the previous pool size should be 1.

I hope you can come up with the reproduction steps for this issue. You can send me e.g. `development.log` or your database file on email or Slack maybe.

libo replied:
```
irb(main):021:0> Pool.delete_all
  Pool Delete All (3.0ms)  DELETE FROM "pools"
=> 3
irb(main):022:0> Event.delete_all
  Event Delete All (1.2ms)  DELETE FROM "events"
=> 3

### Create from web a sell event as first
Event Create (0.4ms)  INSERT INTO "events" ("action", "user_id", "currency_id", "stock_id", "quantity", "price", "commission", "executed_on", "created_at", "updated_at", "sort_column") VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)  [["action", 1], ["user_id", 1], ["currency_id", 1], ["stock_id", 1], ["quantity", 100], ["price", 10.0], ["commission", 0.0], ["executed_on", "2022-01-15"], ["created_at", "2022-01-03 11:46:50.612654"], ["updated_at", "2022-01-03 11:46:50.612654"], ["sort_column", "2022-01-15/1/1641210410.607897"]]
  ↳ app/controllers/events_controller.rb:42:in `create'

  Pool Create (0.4ms)  INSERT INTO "pools" ("event_id", "value", "size") VALUES (?, ?, ?)  [["event_id", 1], ["value", 0.0], ["size", 0]]
  ↳ app/models/pool.rb:33:in `create_event'


irb(main):023:0> Pool.last
  Pool Load (0.1ms)  SELECT "pools".* FROM "pools" ORDER BY "pools"."id" DESC LIMIT ?  [["LIMIT", 1]]
=> #<Pool:0x00007fea329cd988 id: 298, event_id: 1, value: 0.0, size: 0>

```

libo replied:
Also try to add and destroy an event. That creates an exception
```
ActiveRecord::RecordNotFound at /events/1
Couldn't find Event without an ID

in app/models/pool.rb, line 10
   # (Recursively) add a Pool for this event and all after it
    last_event = Event.find(later_event_ids.last)
    create_event(last_event)

```

bquorning replied:
In real life, the first event cannot be a sell. You need to have something in the pool before you can sell from it.

And yes, probably deleting the only event in a pool causes problems. I’ll have a look.

bquorning replied:
I pushed a commit that fixes at least the "add and destroy an event" problem.

libo commented:
An other weird issue in QA.

Try to export and reimport.
<img width="356" alt="Screenshot 2022-01-03 at 12 59 42" src="https://user-images.githubusercontent.com/37512/147928038-0db570b6-0c79-4aff-a29b-d3a6d6a683ee.png">

bquorning committed:
Fix Pool.recalculate

When deleting an event, its pool wouldn't get deleted.
When deleting the last event, an exception was raised.

Both problems are fixed now.

bquorning commented:
> An other weird issue in QA.
> 
> Try to export and reimport. <img alt="Screenshot 2022-01-03 at 12 59 42" width="356" src="https://user-images.githubusercontent.com/37512/147928038-0db570b6-0c79-4aff-a29b-d3a6d6a683ee.png">

I tried, and it seems to work. I need more precise replication steps I’m afraid.

libo commented:
It looks better now. Can you rebase on latest master please so I get also the error messages and the date selection working

bquorning commented:
Rebased. Should I start writing some specs, perhaps?

libo commented on app/models/pool.rb:
```diff
@@ -0,0 +1,60 @@
+    # When deleting an event, its ID is not in "this_and_following_events".
+    Pool.where(event_id: event.id).delete_all if event.destroyed?
+
+    later_event_ids = event.this_and_following_events.pluck(:id)
```
This way of finding the pool items to delete leads to an issue.

Let's say you have an event each month of the year.
You add a 13 event by mistake in Jan
You move it to Dec.

Then this code will look for all the events following this event after we updated its date to dec.
But we actually need to recalculate the pool for all events that were following the event before we changed its date (so Jan - Dec and not Dec only)

bquorning replied:
Good point. We need to find the earliest date from before and after the change.

libo replied:
Maybe move the delete pool associated to the event into a Rails `dependent` scenario and make `recalculate` accept something that can go into a `sort_column` I mean `.where("sort_column >= ?", sort_column)`
